design/design-doc-package-refactor.md

```diff
@@ -0,0 +1,680 @@
+`Stopping`, they must mark it as `Inactive`. Note that if it matches the image
+in its parent `Package` it will likely either by transitioned back to another
+state, or it will be deleted and an identical new `PackageRevision` will be
+created in its place. Users should not attempt to clean up a `PackageRevision`
+that has not been marked as `Inactive` by its parent.
+
+### Phase 1
+
+_Phase 1 should be completed by release of Crossplane v0.13_
+
+Issues to be Opened:
+
+- [ ] Rename `Package` to `LegacyPackage`
+- [ ] Add `Package` and `PackageRevision` types
+- [ ] Implement `Package` and `PackageRevision` controllers
+- [ ] Deprecate `ClusterPackageInstall`, `PackageInstall`, `StackDefinition`,
+  and `LegacyPackage`
+
+### Phase 2
+
+_Phase 2 should be completed by release of Crossplane v0.14_
+
+Issues to be Opened:
+
+- [ ] Deprecate `app-wordpress`
+- [ ] Update `stack-aws-sample` to use new `Package` format and composition
+- [ ] Update `stack-azure-sample` to use new `Package` format and composition
+- [ ] Update `stack-gcp-sample` to use new `Package` format and composition
+
+`ClusterPackageInstall`, `PackageInstall`, `StackDefinition`, and
+`LegacyPackage` and their corresponding controllers are marked as deprecated.
+All stacks should be migrated to use packages that install composition types.
+
+### Phase 3
+
+_Phase 3 should be completed by release of Crossplane v0.15_
+
+Issues to be Opened:
+
+- [ ] Archive `app-wordpress`
+- [ ] Remove `ClusterPackageInstall`, `PackageInstall`, `StackDefinition`, and
+  `LegacyPackage` and their controllers
+
+### Example Manifests
+
+TODO: this section is just experimenting with YAML formats right now
+
+User creates `Package`:
+
+```yaml
+apiVersion: packages.crossplane.io/v1alpha1
+kind: Package
+metadata:
+  name: provider-gcp
+spec:
+  package: crossplane/provider-gcp:v0.10.0
+```
+
+`Package` creates `Job`:
+
+```yaml
+apiVersion: batch/v1
+kind: Job
+metadata:
+  name: provider-gcp
+  namespace: crossplane-system
+  ownerReferences:
+  - apiVersion: packages.crossplane.io/v1alpha1
+    controller: true
+    blockOwnerDeletion: true
+    kind: Package
+    name: provider-gcp
+    uid: d9607e19-f88f-11e6-a518-42010a800195
+spec:
+  template:
+    spec:
+      initContainers:
+      - name: package-copy-to-volume
+        image: crossplane/provider-gcp:v0.10.0
+        cmd: ["cp", "-R", ".registry/", "/ext-pkg/"]
+        volumeMounts:
+        - name: package-contents
+          mountPath: "/ext-pkg"
+      containers:
+      - name: "package-unpack-and-output"
+        args: ["package", "unpack", "--content-dir=/ext-pkg/.registry"]
+        volumeMounts:
+        - name: package-contents
+          mountPath: "/ext-pkg"
+      volumes:
+      - name: package-contents
+        emptyDir: {}
+```
+
+`Package` reads `Job` output and creates `PackageRevision`:
+
+```yaml
+apiVersion: packages.crossplane.io/v1alpha1
+kind: PackageRevision
```

Comment:
Yep, I am familiar with how a `ControllerRevision` works and the `PackageRevision` is not intended to work in the same manner. The reason that we don't just patch the `Package` from a `PackageRevision` here is that the `Package` is intended to be a simple user-facing type. This is opposed to something like a `DaemonSet` where you are actually configuring all of the fields yourself and doing rolling updates. Is your concern with including "revision" in the name? If so, neither the `ControllerRevision` nor the `PackageRevision` is really supposed to be user-facing, so I am skeptical that it will cause a lot of confusion. That being said, `PackageVersion` or any another name in that vein would be fine with me 👍